agent/cache-types/service_checks.go

```diff
@@ -0,0 +1,158 @@
+		}
+	}
+
+	var wait time.Duration
+
+	// Adjust wait based on documented limits and add some jitter: https://www.consul.io/api/features/blocking.html
+	switch wait = reqReal.MaxQueryTime; {
+	case wait == 0*time.Second:
+		wait = 5 * time.Minute
+	case wait > 10*time.Minute:
+		wait = 10 * time.Minute
+	}
+	timeout := time.NewTimer(wait + lib.RandomStagger(wait/16))
+
+	var resp []structs.CheckType
+	var hash string
+
+WATCH_LOOP:
+	for {
+		// Must reset this every loop in case the Watch set is already closed but
+		// hash remains same. In that case we'll need to re-block on ws.Watch()
+		ws := memdb.NewWatchSet()
+
+		svcState := c.Agent.LocalState().ServiceState(reqReal.ServiceID)
+		if svcState == nil {
+			return result, fmt.Errorf("Internal cache failure: service '%s' not in agent state", reqReal.ServiceID)
+		}
+
+		// WatchCh will receive updates on service (de)registrations and check (de)registrations
+		ws.Add(svcState.WatchCh)
+
+		resp = c.Agent.ServiceHTTPChecks(reqReal.ServiceID)
+
+		hash, err = hashChecks(resp)
+		if err != nil {
+			return result, fmt.Errorf("Internal cache failure: %v", err)
+		}
+
+		// Return immediately if the hash is different or the Watch returns true (indicating timeout fired).
+		if lastHash != hash || ws.Watch(timeout.C) {
+			break
+		}
+
+		// Watch returned false indicating a change was detected, loop and repeat
+		// the call to ServiceHTTPChecks to load the new value.
+		// If agent sync is paused it means local state is being bulk-edited e.g. config reload.
+		if syncPauseCh := c.Agent.SyncPausedCh(); syncPauseCh != nil {
+			// Wait for pause to end or for the timeout to elapse.
+			select {
+			case <-syncPauseCh:
+			case <-timeout.C:
+				break WATCH_LOOP
+			}
+		}
+	}
+
+	result.Value = &resp
+
+	// Below is a purely synthetic index to keep the caching happy.
+	if opts.LastResult == nil {
+		result.Index = 1
+		return result, nil
+	}
+
+	result.Index = opts.LastResult.Index
+	if lastHash == "" || hash != lastHash {
+		result.Index += 1
+	}
+	return result, nil
+}
+
+func (c *ServiceHTTPChecks) SupportsBlocking() bool {
+	return true
+}
+
+// ServiceHTTPChecksRequest is the cache.Request implementation for the
+// ServiceHTTPChecks cache type. This is implemented here and not in structs
+// since this is only used for cache-related requests and not forwarded
+// directly to any Consul servers.
+type ServiceHTTPChecksRequest struct {
+	ServiceID     string
+	MinQueryIndex uint64
+	MaxQueryTime  time.Duration
+}
+
+func (s *ServiceHTTPChecksRequest) CacheInfo() cache.RequestInfo {
+	return cache.RequestInfo{
+		Token:      "",
+		Key:        ServiceHTTPChecksName + ":" + s.ServiceID,
+		Datacenter: "",
+		MinIndex:   s.MinQueryIndex,
+		Timeout:    s.MaxQueryTime,
+	}
+}
+
+func hashChecks(checks []structs.CheckType) (string, error) {
+	var b strings.Builder
+	for _, check := range checks {
+		raw, err := hashstructure.Hash(check, nil)
```

Comment:
Also might want to comment that this technically can give a false positive if the array is re-ordered but not changed which you don't care about but will change the hash. If you want to avoid that you can maybe use hash structure set annotation struct tag or just manually XOR the results together instead of concatenating them which will make it order agnostic. I don't think it's that important personally as the cost of a false positive int hat edge case is not high but a comment to show it was considered and might be different for another situation if people copy the pattern might be good.

Reply:
For some reason I thought hashstructure didn't operate directly on slices. Just updated that and added the `set` struct tag.